README.md
FlindersDSpace-csv4itemimport
=============================

Description
-----------
Converts rows in a CSV file (plus optional bitstreams/files)
into a directory and file structure suitable for batch importing as items
into DSpace 3.x using the Simple Archive Format (SAF). The resulting
structure can then be imported into DSpace with a command like:
```
  /path/to/dspace import -a -s OUT_DIR -c COLLECTION_ID -m NEW_MAP_FILE -e USER_EMAIL [--test]
```
where OUT_DIR is the is the top directory of the output structure
created by this program (ie. typically "results/to_import" - defined
by SAF_DEST_FOLDER within bin/csv4itemimport.rb).

**Note that all items will be imported into the same collection.**

The input CSV file is mostly compatible with the
DSpace Batch Metadata Editing CSV file. Similarities include:
- RFC4180 CSV format
- Most columns contain the dublin core metadata fields
- If you want to store multiple values for a given metadata element,
  they must be separated with a double-pipe "||"

Differences include:
- The "id" column is not mandatory and will be ignored.
- The "collection" column is not required and will be ignored.
  All lines in the CSV file will be imported into the *same* collection
  which shall be specified on the "dspace import" command line
  via the -c or --collection option (see the example above).
- A new column "dspace.files" is required if one or more items (ie. CSV lines) has
  corresponding files to be ingested into DSpace via the CSV import
  process. If more than one file needs to be ingested for a single
  item, the list of files must be separated with "||"

In summary, only the following 2 column types shall be recognised:
- dublin core columns starting with "dc."
- the new (but optional) column "dspace.files"

All other column names shall be ignored.

CSV Gotchas
-----------
- Every column must be assigned a name on the first (header) line.
- Any blank columns before the last (right-most) column must either
  be removed or assigned a column name on the first line.
- Beware of single-quote characters (ie. "'") in the first position of
  a cell within a Microsoft Excel XLS or XLSX spreadsheet. These have
  a special meaning to Excel and will be stripped when saving as a
  CSV file. This may not be what you intended.

Application environment
-----------------------
Read the INSTALL file.


Installation
------------
Read the INSTALL file.


CSV file example
----------------
```
  id,collection,dc.contributor.author,dc.description.abstract[en_US],dc.subject[en_US],dc.title[en_US],Status,Editor,dspace.files
  14,123456789/0,Test Contrib Author1,This is the wonderful abstract,subj001||subj002b,Test title #001,Blah..,Blah..,file1.pdf||file2.txt||file3.csv
  15,123456789/0,Test Contrib Author2,This is the wonderful abstract,subj003a||subj004,Test title #002,Blah..,Blah..,
```
Interpretation:
- Columns id, collection, Status and Editor shall be ignored.
- If populated, Dublin Core metadata for the following columns shall be
  ingested into DSpace:
  * dc.contributor.author
  * dc.description.abstract (with language set to "en_US")
  * dc.subject (with language set to "en_US")
  * dc.title (with language set to "en_US")
- In the example above, each item will have 2 values for dc.subject.
- In the example above, there will be 3 bitstreams/files ingested for
  the first item, whereas there will not be any bitstreams/files
  ingested for the second item.


Cleaning the CSV file metadata
------------------------------
If you have good compatibility between the character encoding used in the
input CSV file and the character encoding used by DSpace then no cleaning
of CSV metadata should be needed. In this case you can set the "cleanup_mode"
property to "none" within etc/conf/dublin_core_value_cleanup.yaml and
no cleaning will be performed.

However, I found that the Microsoft Excel spreadsheet (from which the
CSV input file was derived) was created from pasting text from various
sources (eg. Microsoft Word or Adobe Acrobat Reader) and contained several
characters which were unsuitable for our DSpace instance. In particular,
the CSV rows contained metadata text which was outside the printable ASCII
text code range 32-254 (ie. 0x20-0x7e hexadecimal).

### Option 1

The best solution is to use a spreadsheet with the proper character encoding
for your DSpace instance. Below I list a few ways that should allow you to
convert from a Microsoft Excel XSL or XSLX spreadsheet on a PC to a utf-8
CSV file on a Linux host (ie. our DSpace Linux server in my case).

#### Option 1A

- Use Excel to save your XSL/XSLX Windows-1250 encoded spreadsheet as a
  "Unicode Text" file (which is tab delimited and apparently encoded as
  utf-16). As far as I can tell, the quoting (and escaping of quotes) for
  Excel "Unicode Text" is compatible with the CSV specification (RFC 4180)
  so we can replace tabs with commas in a later step.
- Move the Unicode Text file to the Linux DSpace server.
- Convert the encoding of the Unicode Text file to utf-8. Also convert from tab
  delimited to comma delimited. I have successfully carried out these 2
  conversions on the PC with Notepad++ v6.6.7 (using
  *Encoding > Convert to UTF-8* and
  *Search > Replace > Search Mode: Extended; Find what: \t; Replace with: , ;
  Replace All*) but for our environment, I used the following Linux commands.

```
iconv -f UTF16 -t UTF8 input.txt > output_utf8.tab
cat output_utf8.tab |tr '\t' , > output_utf8.csv

# Or as a single command
iconv -f UTF16 -t UTF8 input.txt |tr '\t' , > output_utf8.csv

```

- Configure and run this program to convert the CSV file into DSpace SAF.
  (Before running the program, ensure the "cleanup_mode" property is set
  to "none" in etc/conf/dublin_core_value_cleanup.yaml.)
- Batch import DSpace SAF into DSpace.

#### Option 1B

- Use Excel to save your XSL/XSLX Windows-1250 encoded spreadsheet as a CSV
  file (which will continue to be encoded as Windows-1250). **I found that
  this step already corrupted some non-English characters by replacing them
  with question marks.** During the save as "CSV (Comma delimited)" step in
  Excel 2010, I have tried the following without success:
  * Tools > Web options > Encoding tab > Save this document as: "Unicode (UTF-8)" > OK

- Move the CSV file to the Linux DSpace server.
- Convert the encoding of the CSV file to utf-8. For our environment, I used
  the following Linux command.

```
iconv -f WINDOWS-1250 -t UTF8 input.csv > output_utf8.csv
```

- Configure and run this program to convert the CSV file into DSpace SAF.
  (Before running the program, ensure the "cleanup_mode" property is set
  to "none" in etc/conf/dublin_core_value_cleanup.yaml.)
- Batch import DSpace SAF into DSpace.

#### Option 1C

- Use LibreOffice or OpenOffice (instead of Excel) to save your XSL/XSLX
  Windows-1250 encoded spreadsheet as a CSV with utf-8 encoding. During
  save as "Text (CSV)", click Edit Filter Settings > Character Set:
  "Unicode (UTF-8)" > OK. (I have not tested this.)
- Move the CSV file to the Linux DSpace server.
- Configure and run this program to convert the CSV file into DSpace SAF.
  (Before running the program, ensure the "cleanup_mode" property is set
  to "none" in etc/conf/dublin_core_value_cleanup.yaml.)
- Batch import DSpace SAF into DSpace.


### Option 2

Alternatively, appropriate configuration of etc/conf/dublin_core_value_cleanup.yaml
allowed this issue to be overcome.  In particular:
- the "cleanup_mode" property was set to "fromLookup_toLookupStringWithHtmlCode"
- the string-pairs listed under the "lookup" section allowed
  the 1-byte character specified by the hexadecimal key to be replaced
  by the corresponding string; otherwise
- 1-byte characters between 128-255 (or 0x80-0xff) were replaced with
  their HTML code equivalents (eg. "&amp;#x80;" - "&amp;#xFF;"); otherwise
- the remaining characters were displayed without change

Note that characters 128-159 (or 0x80-0x9f) and some others are
illegal in HTML4, so these are the ones you are most likely to
override in the lookup table.

Although not all characters can be represented by a single byte in most
modern character encodings (eg. utf-8), this work-around allows a
semi-automated solution for the author of the batch import metadata. I
regard this solution as semi-automated rather than fully-automated
because some investigation is needed into:
- which characters used within the CSV file cause a problem for your
  DSpace instance
- a suitable replacement string for such characters



RELEASE_NOTES

FlindersDSpace-csv4itemimport
=============================

Release Notes
-------------

v0.1
----
Initial version. Is able to "clean" CSV text by replacing 1-byte characters
specified in a lookup table with a string. The lookup table is in the
configuration file dublin_core_value_cleanup.yaml.

v0.2
----
- Update data in dublin_core_value_cleanup.yaml
- Allow the DSpace Simple Archive Format item-directory to contain an
  arbitrarily large number of items.
- Add the constant, VERBOSE. If true, the script displays all metadata text
  and shows if any characters are changed via the cleanup process.
- Issue a warning if a single or odd number of '|' are used as this might
  represent a mistyped value separator '||'.
- Issue a warning if filenames referenced more than once.
- Issue an error if any header fields are empty.

v0.3
----
- Improve logging (in VERBOSE mode) to assist character cleanup including
  the detection of illegal HTML4 codes.
- Added new/improved cleanup mode "fromLookup_toLookupStringWithHtmlCode".
  This is the new default cleanup mode. This mode will:
  * substitute any 1-byte character with its associated string from
    the lookup table, otherwise;
  * substitute any 1-byte character with its associated HTML code
    (ie. "&#xNN;") if the byte is 0x80-0xff, otherwise;
  * use the raw character.

v1.0
----
- Document conversion of CSV file to UTF-8 (ie. the preferred cleanup method)
- Changed the default cleanup_mode to 'none'
- Show all the missing files (referenced in the CSV) before quitting

v1.1
----
- Add more warning messages
- Add more robust method for converting Excel spreadsheet to utf-8 CSV file
- Changed CSV filename



lib/item.rb
#--
# Copyright (c) 2014, Flinders University, South Australia. All rights reserved.
# Contributors: eResearch@Flinders, Library, Information Services, Flinders University.
# See the accompanying LICENSE file (or http://opensource.org/licenses/BSD-3-Clause).
#++ 

require 'dublin_core_value'

##############################################################################
# A class to store information regarding a DSpace item for the purposes of
# importing items into DSpace via the Simple Archive Format.
class Item
  # In a single CSV column, use this double-character delimiter to separate multiple values
  VALUE_DELIMITER = '||'
  # Beware of single (or odd number of) delimiter characters eg. '|'
  BAD_VALUE_DELIMITER = VALUE_DELIMITER[/^./]
  NEWLINE = "\n"

  XML_HEADER_LINES = [
    "<dublin_core>",
  ]

  XML_FOOTER_LINES = [
    "</dublin_core>",
  ]

  attr_reader :filenames

  ############################################################################
  # Create an Item object.
  #
  # If an item_hash value is nil or a sub-field of the value (by splitting
  # with VALUE_DELIMITER) is empty, we will not create a DublinCoreValue
  # object for that entity.
  def initialize(item_hash, filenames_str='')
    @item_hash = item_hash	# Hash of DC names (CSV column names) as keys with corresponding values
    @dc_list = []		# List of DublinCoreValue objects
    @filenames = []		# List of filenames (bitstreams) to be ingested

    @item_hash.each{|dc_name,dc_values|
      next unless dc_values

      # If dc_values string contains VALUE_DELIMITER, then we need to
      # split the string into fields (based on the delimiter) and
      # create a DublinCoreValue for each (value) field.
      has_bad_delim = false
      dc_values.split(VALUE_DELIMITER).each{|dc_value|
        stripped_dc_value = dc_value.strip
        next if stripped_dc_value.empty?
        d = DublinCoreValue.new(dc_name, stripped_dc_value)
        has_bad_delim ||= stripped_dc_value.include?(BAD_VALUE_DELIMITER)
        @dc_list << d
      }
      if has_bad_delim
        STDERR.puts <<-MSG_BAD_VALUE_DELIMITER.gsub(/^\t*/, '')
		WARNING: Possible bad delimiter (odd number of \"#{BAD_VALUE_DELIMITER}\") instead of \"#{VALUE_DELIMITER}\" in:
		  \"#{dc_values.strip}\"
		  for CSV column: \"#{dc_name}\"
        MSG_BAD_VALUE_DELIMITER
      end
    }

    if filenames_str
      # BAD_VALUE_DELIMITER here will be found later (as file-not-found)
      filenames_str.split(VALUE_DELIMITER).each{|fname|
        stripped_fname = fname.strip
        @filenames << stripped_fname unless stripped_fname.empty?
      }
    end
  end

  ############################################################################
  # Convert this object into a string representing the whole of a
  # Dublin Core XML file compatible with the DSpace Simple Archive Format.
  def to_xml
    puts "#{self.class}.#{__method__} invoked for object_id #{object_id}" if DublinCoreValue::VERBOSE
    s = XML_HEADER_LINES.join(NEWLINE)
    s += NEWLINE + @dc_list.sort.collect{|d| "  #{d}"}.join(NEWLINE)
    s += NEWLINE + XML_FOOTER_LINES.join(NEWLINE)
    s
  end

  ############################################################################
  alias to_s to_xml

end



lib/items.rb
#--
# Copyright (c) 2014, Flinders University, South Australia. All rights reserved.
# Contributors: eResearch@Flinders, Library, Information Services, Flinders University.
# See the accompanying LICENSE file (or http://opensource.org/licenses/BSD-3-Clause).
#++ 

require 'faster_csv'
require 'item'
require 'object_extra'
require 'fileutils'

##############################################################################
# A class to store information regarding a group of DSpace items for the
# purpose of importing that group into DSpace using the Simple Archive Format.
class Items
  # SAF item-directory shall be named as CHILD_DIR_PREFIX plus a zero-padded
  # numeric suffix of MIN_ITEM_NUMBER_WIDTH digits (or more digits if needed).
  CHILD_DIR_PREFIX = 'item'
  MIN_ITEM_NUMBER_WIDTH = 4
  FILENAMES_CSV_COLUMN = 'dspace.files'

  DC_FILENAME = 'dublin_core.xml'
  CONTENTS_FILENAME = 'contents'

  attr_reader :parent_dir_name

  ############################################################################
  # Create an aggregation of items.
  def initialize(parent_dir_name, bitstream_source_dir, cleanup_config_filename)
    @parent_dir_name = parent_dir_name
    @bitstream_source_dir = bitstream_source_dir
    @cleanup_config_filename = cleanup_config_filename
    DublinCoreValue.cleanup_config_filename = @cleanup_config_filename
    @items = []
  end

  ############################################################################
  # Return the number of items in this object.
  def length
    @items.length
  end

  ############################################################################
  # Load/populate this aggregation of items from the specified CSV file.
  def load_csv(fname, faster_csv_options={})
    opts = {
      :col_sep => ',',
      # It is not advisible to override the values below with those
      # from faster_csv_options (as this method assume these values)
      :headers => true,
      :header_converters => nil,	# CSV header names are a string (not a symbol)
    }.merge!(faster_csv_options)

    count = 0
    count_missing_column4files = 0
    FasterCSV.foreach(fname, opts) {|line|
      count += 1
      puts "\n#{count} <<#{line.to_s.chomp}>>" if MDEBUG.include?(__method__)

      item_hash = {}
      line.each{|key, value|
        unless key
          STDERR.puts "\nERROR: One of the CSV header fields is empty (ie. nil)."
          STDERR.puts "Header fields: #{line.headers.inspect}"
          exit 7
        end
        item_hash[key] = value if key.match(/^dc\./) 
      }

      count_missing_column4files += 1 unless line[FILENAMES_CSV_COLUMN]
      STDERR.puts "WARNING: All metadata fields are empty in record #{count}" if item_hash.all?{|k,v| v==nil}
      @items << Item.new(item_hash, line[FILENAMES_CSV_COLUMN]) unless item_hash.empty?
    }
    STDERR.puts "WARNING: CSV column '#{FILENAMES_CSV_COLUMN}' does not contain DSpace filenames in #{count_missing_column4files} records" unless count_missing_column4files == 0
    verify_all_files_exist
    warn_if_filenames_repeated
  end

  ############################################################################
  # Verify that all files (referenced in the CSV) exist. If any files
  # are missing, show all the missing files before quitting (rather
  # than only one at a time as per the populate_folder() method).
  def verify_all_files_exist
    will_quit = false
    @items.each{|item|
      item.filenames.each{|fname|
        fpath = "#{@bitstream_source_dir}/#{fname}"
        unless File.file?(fpath)
          STDERR.puts "ERROR: File '#{fpath}' not found."
          will_quit = true
        end
      }
    }
    exit 6 if will_quit
  end

  ############################################################################
  # Issue warning if any filenames have been referenced more than once.
  def warn_if_filenames_repeated
    all_files = @items.inject([]){|a,item| a += item.filenames}
    dup_files = all_files.find_all{|f| all_files.count(f) > 1}.uniq

    if dup_files.length > 0
      STDERR.puts "\nWARNING: A reference to one or more 'dspace.files' has been repeated."
      dup_files.sort.each{|f| STDERR.puts "- 'dspace.files' filename:  #{f}"}
    end
  end

  ############################################################################
  # Populate the specified folder with a structure representing this
  # aggregation of items.
  def populate_folder
    # Make parent dir
    if File.exists? @parent_dir_name
      STDERR.puts "ERROR: Cannot create directory '#{@parent_dir_name}'."
      STDERR.puts "Directory or file already exists"
      exit 5
    else
      FileUtils.mkdir_p @parent_dir_name
    end

    # Make and populate child dirs
    width = [MIN_ITEM_NUMBER_WIDTH, (@items.length - 1).to_s.length].max
    print_format = "%s/%s%0#{width}d"

    @items.each_with_index{|item, i|
      child_dir = sprintf print_format, @parent_dir_name, CHILD_DIR_PREFIX, i
      child_dc_path = "#{child_dir}/#{DC_FILENAME}"
      child_contents_path = "#{child_dir}/#{CONTENTS_FILENAME}"
      FileUtils.mkdir child_dir

      # Write item.to_xml to child_dc_path
      File.write_string(child_dc_path, item.to_xml)

      # Write to child_contents_path
      if item.filenames.empty?
        # Write an empty contents file
        FileUtils.touch child_contents_path
      else
        item.filenames.each{|fname|
          # Copy each file into the child dir
          fpath = "#{@bitstream_source_dir}/#{fname}"
          unless File.file?(fpath)
            STDERR.puts "ERROR: File '#{fpath}' not found."
            exit 6
          end
          FileUtils.copy(fpath, child_dir)
        }
        # Write the filenames to the contents file
        filenames_str = item.filenames.join(Item::NEWLINE) + Item::NEWLINE
        File.write_string(child_contents_path, filenames_str)
      end
    }
  end

  ############################################################################
  # Returns a string which contains a human-friendly portrayal of the object.
  def inspect
    "#{@parent_dir_name} -- #{@items.inspect}"
  end

end



lib/dublin_core_value.rb
#--
# Copyright (c) 2014, Flinders University, South Australia. All rights reserved.
# Contributors: eResearch@Flinders, Library, Information Services, Flinders University.
# See the accompanying LICENSE file (or http://opensource.org/licenses/BSD-3-Clause).
#++ 

require 'yaml'
require 'cgi'

##############################################################################
# A class to store information regarding a Dublin Core metadata line/object
# for use with the DSpace Simple Archive Format.
class DublinCoreValue
  include Comparable

  VERBOSE = true

  DEFAULT_CLEANUP_CONFIG_FILENAME = "../etc/conf/dublin_core_value_cleanup.yaml"

  NEWLINE = "\n"

  # Cleanup modes which are permitted within the YAML config file.
  CLEANUP_MODES = %w(
    fromAbove7f_toHtmlCode
    fromLookup_toHtmlCode
    fromLookup_toLookupString
    fromLookup_toLookupStringWithHtmlCode
    none
  )

  # Cleanup modes which use the list of hash-keys specified in the YAML config file.
  CLEANUP_MODES_WITH_KEY = CLEANUP_MODES.inject([]){|a,mode| a << mode if mode.match(/^fromLookup/); a}

  # Cleanup modes which use the list of hash-values specified in the YAML config file.
  CLEANUP_MODES_WITH_VALUE = CLEANUP_MODES.inject([]){|a,mode| a << mode if mode.match(/toLookupString/); a}

  # HTML-code format-string for printf(). Eg. "&#xA9;"
  HTML_CODE_FORMAT = "&#x%X;"

  # Illegal HTML4 character encodings (in hexadecimal) are:
  # 00-1f (except 09, 0a, 0d); 7f; 80-9f; d800-dfff
  HTML_ILLEGAL_CODE_RANGES = [
    0x00..0x08,
    0x0b..0x0c,
    0x0e..0x0f,
    0x7f..0x7f,
    0x80..0x9f,
    0xd800..0xdfff,
  ]

  @@cleanup_config_yaml_filename = DEFAULT_CLEANUP_CONFIG_FILENAME
  @@cleanup_properties = nil
  @@cleanup_lookup = nil

  attr_reader :name, :value

  ############################################################################
  # Create a single Dublin Core Value object.
  #
  # Argument 'name' has format: 'dc.element_name.qualifier_name[language_abbreviation]'
  #
  # where
  # - the '.qualifier_name' is optional.
  # - the '[language_abbreviation]' is optional.
  # For example, the following are all valid for this program.
  # - dc.identifier.uri[en_US]
  # - dc.identifier[en_US]
  # - dc.identifier.uri
  # - dc.identifier
  #
  # Argument 'value' is the corresponding metadata text.
  def initialize(name, value)
    @name = name
    @value = value

    parts = @name.split('.')
    unless parts[0].match('^dc$')
      STDERR.puts "ERROR: '#{@name}' does not represent a Dublin Core element (must start with 'dc.')."
      exit 2
    end

    unless [2,3].include?(parts.length)
      STDERR.puts "ERROR: '#{@name}' does not represent a Dublin Core element."
      STDERR.puts "It must contain 1 or 2 dots."
      exit 3
    end

    # Extract element, qualifier & language
    subparts = parts.last.split(/[\[\]]/)

    @lang = subparts.length == 2 ? subparts[1] : nil
    @element = nil
    @qualifier = nil
    if parts.length == 3
      @element = parts[1]
      @qualifier = subparts[0]
    else  # parts.length == 2
      @element = subparts[0]
    end

    unless @@cleanup_properties
      self.class.load_cleanup_parameters unless @@cleanup_properties
      puts "#{self.class} cleanup mode:     #{@@cleanup_properties['cleanup_mode']}"
    end
  end

  ############################################################################
  # Comparison operator to allow objects of this class to be sorted, etc.
  def <=>(other)
    @name == other.name ? @value <=> other.value : @name <=> other.name
  end

  ############################################################################
  # Convert this object to an XML element (compatible with the DSpace
  # Simple Archive Format).
  def to_xml
    s = []
    s << "<dcvalue element=\"#{@element}\" "
    s << "qualifier=\"#{@qualifier}\" " if @qualifier
    s << "language=\"#{@lang}\" " if @lang
    s << ">#{to_s_value_cleanup}</dcvalue>"
    s.join
  end

  ############################################################################
  # Load parameters from the YAML cleanup config file.
  def self.load_cleanup_parameters
    fname = @@cleanup_config_yaml_filename
    nl = NEWLINE

    begin
      conf_hash = YAML.load_file(fname)

    rescue Exception => ex
      msgs = []
      msgs << "### #{ex}#{nl}"
      msgs << "### YAML file '#{fname}' cannot be loaded.#{nl}"
      msgs << "### WARNING: No #{self} text cleanup will be performed.#{nl}"
      self.set_default_cleanup_properties(msgs.join)
      return
    end

    begin
      @@cleanup_properties = conf_hash['properties']
      raise "'cleanup_mode' is not specified in file #{fname}" unless @@cleanup_properties['cleanup_mode']
      raise "Invalid 'cleanup_mode' specified in file #{fname}" unless CLEANUP_MODES.include?(@@cleanup_properties['cleanup_mode'])

    rescue Exception => ex
      msgs = []
      msgs << "### #{ex}#{nl}"
      msgs << "### WARNING: No #{self} text cleanup will be performed.#{nl}"
      self.set_default_cleanup_properties(msgs.join)
      return
    end

    if CLEANUP_MODES_WITH_KEY.include?(@@cleanup_properties['cleanup_mode'])
      begin
        @@cleanup_lookup = conf_hash['lookup']
        raise "'lookup' is not specified in file #{fname}" unless @@cleanup_lookup.class == Hash

        keys_ok = @@cleanup_lookup.all?{|(key,value)| key.kind_of?(Integer) && key >= 0 && key <= 0xff}
        raise "Not all lookup keys are integers in the range 0-255 (0x00-0xff) inclusive." unless keys_ok

        value_strings_ok = @@cleanup_lookup.all?{|(key,value)| value.kind_of?(String)}	# Or just check if the object has a to_s method?
        raise "Not all lookup values are strings" if CLEANUP_MODES_WITH_VALUE.include?(@@cleanup_properties['cleanup_mode']) && !value_strings_ok

      rescue Exception => ex
        msgs = []
        msgs << "### #{ex}#{nl}"
        msgs << "### WARNING: No #{self} text cleanup will be performed.#{nl}"
        self.set_default_cleanup_properties(msgs.join)
        return
      end
    end
  end

  ############################################################################
  # Set default cleanup properties.
  def self.set_default_cleanup_properties(msg=nil)
    STDERR.puts msg if msg
    @@cleanup_properties = {'cleanup_mode' => 'none'}
  end

  ############################################################################
  # Configure the cleanup-config filename.
  def self.cleanup_config_filename=(filename)
    @@cleanup_config_yaml_filename = filename
  end

  ############################################################################
  # Return true if integer argument is an illegal HTML code, else return false
  def self.html_illegal_code?(int)
    HTML_ILLEGAL_CODE_RANGES.each{|rng| return true if rng.include?(int)}
    false
  end

  ############################################################################
  # Convert 'value' to a string by:
  # - escaping HTML special characters
  # - applying cleanup character conversions specified in the cleanup-config file
  def to_s_value_cleanup
    puts __method__ if VERBOSE
    fmt = HTML_CODE_FORMAT
    hvalue = CGI::escapeHTML(@value)

    case @@cleanup_properties['cleanup_mode']

    # Replace each byte from 128-255 (0x80-0xff) with its HTML code. Eg. "&#xA9;"
    when 'fromAbove7f_toHtmlCode'
      new_chars = []
      hvalue.each_byte{|b| new_chars << (b >= 0x80 && b <= 0xff ? sprintf(fmt, b) : b.chr) }
      new_chars.join

    # Use lookup table to replace each byte with its HTML code. Eg. "&#xA9;"
    when 'fromLookup_toHtmlCode'
      new_chars = []
      hvalue.each_byte{|b| new_chars << (@@cleanup_lookup[b] ? sprintf(fmt, b) : b.chr) }
      new_chars.join

    # Use lookup table to replace each byte with associated string (also from the table)
    when 'fromLookup_toLookupString'
      new_chars = []
      hvalue.each_byte{|b|
        if @@cleanup_lookup[b]
          new_char = @@cleanup_lookup[b]
          msg = ' %%% Updated %%%'
        else
          new_char = b.chr
          msg = b < 0x80 ? '' : ' %%% WARNING: Non-ASCII character %%%'
        end
        printf("=== CLEANUP (%3d, 0x%2x)  %s  %s%s\n", b, b, b.chr, new_char, msg) if VERBOSE
        new_chars << new_char
      }
      new_chars.join

    # Rules:
    # - Use lookup table to replace each byte with associated string (also from the table); else
    # - Replace each byte from 128-255 (0x80-0xff) with its HTML code (eg. "&#xA9;"); else
    # - Use the raw character
    # This allows us to override illegal HTML character encodings (ie. hex 00-1f (except 09, 0a, 0d);
    # 7f; 80-9f; d800-dfff) & anything else at our discretion.
    when 'fromLookup_toLookupStringWithHtmlCode'
      new_chars = []
      hvalue.each_byte{|b|
        if @@cleanup_lookup[b]
          new_char = @@cleanup_lookup[b]
          msg = ' %%% Updated with lookup-string'
        else
          if b >= 0x80 && b <= 0xff
            new_char = sprintf(fmt, b)
            msg = ' %%% Updated with HTML-code'
          else
            new_char = b.chr
            msg = ''
          end
          msg += ' %%% WARNING: Illegal HTML-code' if self.class.html_illegal_code?(b)
        end
        printf("=== CLEANUP (%3d, 0x%2x)  %s  %s%s\n", b, b, b.chr, new_char, msg) if VERBOSE
        new_chars << new_char
      }
      new_chars.join

    # Do not apply any filter
    when 'none'
      hvalue

    # Should never reach here!
    else
      STD.puts "ERROR: Unrecognised cleanup mode #{@@cleanup_properties['cleanup_mode']}"
      exit 4
    end

  end

  ############################################################################
  alias to_s to_xml

end

